src/plugins/embeddable/public/lib/containers/container.ts

```diff
@@ -195,6 +197,32 @@ export abstract class Container<
     });
   }
 
+  public async getExplicitInputIsEqual(lastInput: TContainerInput) {
+    const { panels: lastPanels, ...restOfLastInput } = lastInput;
+    const { panels: currentPanels, ...restOfCurrentInput } = this.getInput();
+    const otherInputIsEqual = isEqual(restOfLastInput, restOfCurrentInput);
+    if (!otherInputIsEqual) return false;
+
+    const embeddableIdsA = Object.keys(lastPanels);
+    const embeddableIdsB = Object.keys(currentPanels);
+    if (
+      embeddableIdsA.length !== embeddableIdsB.length ||
+      xor(embeddableIdsA, embeddableIdsB).length > 0
+    ) {
+      return false;
+    }
+    // embeddable ids are equal so let's compare individual panels.
+    for (const id of embeddableIdsA) {
+      const currentEmbeddable = await this.untilEmbeddableLoaded(id);
+      const lastPanelInput = lastPanels[id].explicitInput;
+      if (isErrorEmbeddable(currentEmbeddable)) continue;
+      if (await currentEmbeddable.getExplicitInputIsEqual(lastPanelInput)) {
```

Comment:
Should it be? 

```suggestion
      if (!(await currentEmbeddable.getExplicitInputIsEqual(lastPanelInput))) {
```

Or am I missing something

Reply:
Ah yes, good catch, I've committed this.

The implementation in container.ts isn't used yet - which is why that obvious issue with the logic wasn't causing problems. It's there mostly as a TODO for now. See the TODO comment `src/plugins/dashboard/public/application/lib/diff_dashboard_state.ts` Line 32